### PR TITLE
#1665. Improve Package Selection

### PR DESCRIPTION
**Pull request by mike-winberry**
## Is your feature request related to a problem? Please describe.
Our package selection workflow in the UI is a bit slow due to searching the entire home directory at one time.

## Describe the solution you'd like
Recommend the following changes:

### For 'init' package deployment
- [x] in the selection window, only show init packages that exist in the three places that Zarf searches for them: 
  - Current Working Directory
  -  Zarf Binary Directory
  - Home `.zarf-cache` folder
  - Only show init packages that match the current build version. 

### Regular Package Deployment
For 'regular' packages -  the UX flow proceeds in two parts:
- [x] Search # 1. 
     - Search and show packages located in the current working directory w/o subdirectories (matching the tab behavior on `zarf package deploy`)
    - If packages are found show in the table and add a expand search button to table that executes search # 2
    - If no packages are found - show no package found error message (gives the user option to cancel deployment or proceed to part # 2
- [x] search part # 2
    - Search and show packages located in the entire home directory
    - Exclude trash 
    - Exclude hidden folders

- [x] For both workflows, increase the height of the package selection table to show more packages at once

## Design File
https://www.figma.com/file/MUItIzpzLBLuIyt225Bwgl/Zarf-Web-Ui?type=design&node-id=1544%3A56348&t=JeB5xVvOb7lcP6ZB-1

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by mike-winberry** [comment]
OS File Picker will not work since it is really an upload input and does not reveal the local systems file structure (https://html.spec.whatwg.org/multipage/input.html#fakepath-srsly) leaving options:
- Education
- Roll your own file explorer
- Be a little slow.

**Review by Madeline-UX — COMMENTED** [review]
Reviewed the solution that @mike-winberry created for the system browser proposed by the team. 

Given the user of this feature is Ashton we do not assume this persona is competent with CLI tool. After reviewing the issue, I do not agree with or approve of this UX flow. 
Reasons:

- unnecessarily complex - User clicking through a file system requires more cognitive load than waiting for a search to populate. User must remember or spend time searching through folders
- UX flow of searching file system may imply to some users that a file is being uploaded.
- UI is not consistent with system browser - can lead to confusion for user. 
- Does not explain or provide reason for why Zarf only searches local system
- Why would we require a user to look for something we can just show them.

@Racer159  Requesting this issue be put on hold until UX design is addressed: 

Options to explore
1. Keep MVP as is until we can gather data on observable user behavior (optimize for speed)
2. Error message that asks if user would like to search entire system post a local search
3. Error message telling user to move file to working directory and hope they know how

**Comment by Racer159** [comment]
> Reviewed the solution that @mike-winberry created for the system browser proposed by the team.
> 
> Given the user of this feature is Ashton we do not assume this persona is competent with CLI tool. After reviewing the issue, I do not agree with or approve of this UX flow. Reasons:
> 
>     * unnecessarily complex - User clicking through a file system requires more cognitive load than waiting for a search to populate. User must remember or spend time searching through folders
> 
>     * UX flow of searching file system may imply to some users that a file is being uploaded.
> 
>     * UI is not consistent with system browser - can lead to confusion for user.
> 
>     * Does not explain or provide reason for why Zarf only searches local system
> 
>     * Why would we require a user to look for something we can just show them.
> 
> 
> @Racer159 Requesting this issue be put on hold until UX design is addressed:
> 
> Options to explore
> 
>     1. Keep MVP as is until we can gather data on observable user behavior (optimize for speed)
> 
>     2. Error message that asks if user would like to search entire system post a local search
> 
>     3. Error message telling user to move file to working directory and hope they know how

Pushed it back - of these I think 1 and 2 are maybe the best - testing it again it is still a bit slow even with local dir packages (I have more than the average bear but it takes several seconds to return the package information on my system).

**Comment by mike-winberry** [comment]
https://github.com/defenseunicorns/zarf/assets/86802655/f1c18a53-e24c-42ee-9a3c-1f488ff22c11


https://github.com/defenseunicorns/zarf/assets/86802655/48fdd773-be91-446f-b200-768aab9d951c


https://github.com/defenseunicorns/zarf/assets/86802655/4efd8822-6bb3-4de0-8da2-c5c13db88e66

File: src/internal/api/packages/find.go [hunk 1]
```diff
@@ -5,52 +5,177 @@
...
+	w.Header().Set("Connection", "keep-alive")
+
+	done := make(chan bool)
+	go func() {
+		// stream init packages in the execution directory
+		if execDir, err := os.Getwd(); err == nil {
+			streamDirPackages(execDir, currentInitPattern, w)
+		} else {
+			streamError(err, w)
+		}
```
Comment (Racer159) [inline src/internal/api/packages/find.go:69]: This doesn't get the binary location. See `GetFinalExecutablePath()` in `src/pkg/utils/io.go`

**Review by Racer159 — CHANGES_REQUESTED** [review]
Overall much better UX, just a few small things

File: src/pkg/utils/io.go [hunk 2]
```diff
@@ -196,13 +190,45 @@ func RecursiveFileList(dir string, pattern *regexp.Regexp, skipPermission bool,
 			} else {
 				files = append(files, path)
 			}
+			// Skip hidden directories
+		} else if IsHidden(d.Name()) {
+			return filepath.SkipDir
 		}
```
[the remark this answers is not in the document]
Comment (mike-winberry) [reply src/pkg/utils/io.go:196]: I can add that back in with a flag. I wasn't sure b/c that was the original behavior prior to the "new" search functionality in find that I added to make it work before this PR.

File: src/internal/api/packages/find.go (hunk 1 above)
Comment (mike-winberry) [reply src/internal/api/packages/find.go:69]: Yep I missed that when I was moving things around.

File: src/internal/packager/template/yaml.go [hunk 3]
```diff
@@ -15,7 +15,7 @@ import (
 func ProcessYamlFilesInPath(path string, component types.ZarfComponent, values Values) ([]string, error) {
 	// Only pull in yml and yaml files
 	pattern := regexp.MustCompile(`(?mi)\.ya?ml$`)
-	manifests, _ := utils.RecursiveFileList(path, pattern, false, true)
+	manifests, _ := utils.RecursiveFileList(path, pattern, true)
```
Comment (Racer159) [inline src/internal/packager/template/yaml.go:18]:
```suggestion
	manifests, _ := utils.RecursiveFileList(path, pattern, false)
```

File: src/pkg/packager/create.go [hunk 4]
```diff
@@ -302,7 +302,7 @@ func (p *Packager) getFilesToSBOM(component types.ZarfComponent) (*types.Compone
 
 	appendSBOMFiles := func(path string) {
 		if utils.IsDir(path) {
-			files, _ := utils.RecursiveFileList(path, nil, false, true)
+			files, _ := utils.RecursiveFileList(path, nil, true)
```
Comment (Racer159) [inline src/pkg/packager/create.go:305]:
```suggestion
			files, _ := utils.RecursiveFileList(path, nil, false)
```

File: src/pkg/packager/deploy.go [hunk 5]
```diff
@@ -317,7 +317,7 @@ func (p *Packager) processComponentFiles(component types.ZarfComponent, pkgLocat
 
 		fileList := []string{}
 		if utils.IsDir(fileLocation) {
-			files, _ := utils.RecursiveFileList(fileLocation, nil, false, true)
+			files, _ := utils.RecursiveFileList(fileLocation, nil, true)
```
Comment (Racer159) [inline src/pkg/packager/deploy.go:320]:
```suggestion
			files, _ := utils.RecursiveFileList(fileLocation, nil, false)
```

**Review by Racer159 — CHANGES_REQUESTED** [review]
skipHidden should be false on Files (was likely a bug if it wasn't before)

**Review by Madeline-UX — DISMISSED** [review]

**Review by Racer159 — APPROVED** [review]
lgtm

**Review by Racer159 — APPROVED** [review]

**Pull request merged**